File: Base/순열조합부분집합/순열.py
from pprint import pprint


def PERM(arr, r):
    result = []
    used = [0 for _ in range(len(arr))]

    def perm(k, chosen, used):
        if k == r:
            # 파이썬은 포인터값을 저장하기 때문에 참조 원본이 바뀌면 복사해서 넣어주어야 함
            result.append(chosen[::])
            return
        for i in range(len(arr)):
            if used[i]:
                continue
            chosen.append(arr[i])
            used[i] = 1
            perm(k+1, chosen, used)
            used[i] = 0
            chosen.pop()
    perm(0, [], used)
    return result


result = PERM('ABCDE', 3)
pprint(result)
pprint(str(len(result)) + "개")

File: 자스베이스/순열조합/순열조합/순열.js
const perm = (k, N, used, choice, numbers) => {
    if (k === N) {
        console.log(choice.join(" "));
        return;
    }
    for (let i = 0; i < numbers.length; i++) {
        if (used & (1 << i)) continue;
        choice.push(numbers[i]);
        perm(k + 1, N, used | (1 << i), choice, numbers);
        choice.pop();
    }
};

perm(0, 2, 0, [], [1, 2, 3]);
